tracer/src/Datadog.Trace/Ci/Agent/CIAgentlessWriter.cs

```diff
@@ -0,0 +1,217 @@
+        public Task FlushAndCloseAsync()
+        {
+            _eventQueue.CompleteAdding();
+            _flushDelayEvent.Set();
+            return _flushTaskCompletionSource.Task;
+        }
+
+        public Task FlushTracesAsync()
+        {
+            var wme = new WatermarkEvent();
+            if (_eventQueue.IsAddingCompleted)
+            {
+                return Task.CompletedTask;
+            }
+
+            try
+            {
+                _eventQueue.Add(wme);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error Writing event in a queue.");
+                return Task.FromException(ex);
+            }
+
+            _flushDelayEvent.Set();
+            return wme.CompletionSource.Task;
+        }
+
+        public Task<bool> Ping()
+        {
+            return Task.FromResult(true);
+        }
+
+        public void WriteTrace(ArraySegment<Span> trace)
+        {
+            // Transform spans to events
+
+            for (var i = trace.Offset; i < trace.Count; i++)
+            {
+                if (trace.Array[i].Type == SpanTypes.Test)
+                {
+                    WriteEvent(new TestEvent(trace.Array[i]));
+                }
+                else
+                {
+                    WriteEvent(new SpanEvent(trace.Array[i]));
+                }
+            }
+        }
+
+        private static async Task InternalFlushEventsAsync(object state)
+        {
+            var writer = (CIAgentlessWriter)state;
+            var eventQueue = writer._eventQueue;
+            var completionSource = writer._flushTaskCompletionSource;
+            var flushDelayEvent = writer._flushDelayEvent;
+            var ciTestCycleBuffer = writer._ciTestCycleBuffer;
+
+            Log.Debug("CIAgentlessWriter:: InternalFlushEventsAsync/ Starting FlushEventsAsync loop");
+
+            while (!eventQueue.IsCompleted)
+            {
+                TaskCompletionSource<bool> watermarkCompletion = null;
+
+                try
+                {
+                    // Retrieve events from the queue and add them to the respective buffer.
+                    while (eventQueue.TryTake(out var item))
+                    {
+                        if (item is WatermarkEvent watermarkEvent)
+                        {
+                            // Flush operation.
+                            // We get the completion source and exit this loop
+                            // to flush buffers (in case there's any event)
+                            watermarkCompletion = watermarkEvent.CompletionSource;
+                            break;
+                        }
+                        else if (ciTestCycleBuffer.CanProcessEvent(item))
+                        {
+                            // The CITestCycle endpoint can process this event, we try to add it to the buffer.
+                            if (!ciTestCycleBuffer.TryProcessEvent(item) && ciTestCycleBuffer.HasEvents)
```

Comment:
If an item isn't processed, do we drop it? It seems like we don't try adding an event a second time

Reply:
oh you're right, I think changing the `if` with a `while` should do the job.

Reply:
done in [3f15402](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/3f15402afc5a8d40a27ef2211a2197589be4a9ed)